src/main/scala/org/broadinstitute/hail/expr/TypeTag.scala

```diff
@@ -0,0 +1,23 @@
+package org.broadinstitute.hail.expr
+
+object TypeTag {
+  def funTypeFromArgs(args: Seq[BaseType]): TypeTag = args match {
+    case Seq(a) => UnaryFunType(a)
+    case Seq(a, b) => BinaryFunType(a, b)
+    case Seq(a, b, c) => Arity3FunType(a, b, c)
+    case Seq(a, b, c, d) => Arity4FunType(a, b, c, d)
+    case _ => throw new RuntimeException(s"Unsupported number of arguments: ${ args.length }")
+  }
+}
+
+sealed trait TypeTag
+
+sealed case class UnaryMethodType(x: BaseType) extends TypeTag
+
+sealed case class UnaryFunType(x: BaseType) extends TypeTag
+
+sealed case class BinaryFunType(x: BaseType, y: BaseType) extends TypeTag
+
+sealed case class Arity3FunType(x: BaseType, y: BaseType, z: BaseType) extends TypeTag
+
+sealed case class Arity4FunType(x: BaseType, y: BaseType, z: BaseType, a: BaseType) extends TypeTag
```

Comment:
Why don't the `FunType`s take the argument types as a `Seq`?  Then you don't need the different arities and you get two versions, `UnaryMethodType` (fields) and `FunType`.  From the perspective, I'd treat `this` for a Method as the first argument.